If the first numeric index of an array is negative let the next one stay negative

pmmaga opened the pull request:
The current behavior of `array_fill(-10, 3, true)` - although [documented](http://php.net/manual/en/function.array-fill.php) - is [unexpected](https://3v4l.org/CtPsu).
This same behavior can also be seen in [array declarations with an explicit negative index](https://3v4l.org/Ytb2J).

IMHO, this sits in a limbo between a bugfix and a language change. Or, in other words, it seems to me that the only reason why this would not be a bug would be the fact that it is documented.

Let me know what you think.

Link to the RFC: https://wiki.php.net/rfc/negative_array_index

krakjoe commented:
I'm not sure how to label this: Something that is documented, and that works the same in every released version of PHP (and hhvm) can't really be described as unexpected, or buggy. It's surprising at the most ...

I think this needs an RFC to resolve it, so I'm going to label it as such ... at the very least it requires an internals discussion.

pmmaga commented:
Agreed. I'll start a discussion on internals later today. Thanks for the feedback!

andrewnester commented:
sorry guys, I didn't see this PR before I sent my one, but I have alternative approach to fix this.
https://github.com/php/php-src/pull/2384

sgolemon commented:
I don't think it's a thing that needs fixing (although it is wtfaf).  At most, I'd say deprecate this method and introduce a new one, but even that feels heavy handed given how positively ancient this function is.

pmmaga commented:
@sgolemon, thanks for the feedback.

If you check the implementation, you'll see that this change wouldn't be limited to the array_fill function but to all arrays where the first key is negative. So, in the scope of this implementation, deprecating a function/introducing a new one wouldn't really make sense.

andrewnester commented:
@pmmaga I am wondering how it will influence existing code which relies on previous version of array indexes?

pmmaga commented:
@andrewnester, yes, it would introduce a BC break in those cases.

nikic commented:
Looks like the latest changes broke the build completely...

pmmaga commented:
It did indeed. I didn't have the time to fix this today so I added the [WIP] to the title until I get it right again.

pmmaga commented:
Ok, I've managed to fix the issues and it's all green now.

Majkl578 commented:
This definitely should **not** target PHP 7.2 as it breaks backward compatibility.

ciaranmcnulty commented:
'Documented but unexpected' doesn't make any sense to me?

 Changing documented behaviour is a clear BC break so this should target `8.0` IMO

ciaranmcnulty commented:
I'd rather see an extra param to `array_fill` then change the default in `8.0`, FWIW

nicolas-grekas commented:
see https://wiki.php.net/rfc/releaseprocess
it's pretty clear this cannot into 7.2

tpunt commented:
I find the current behaviour of `array_fill` unintuitive and quite useless. But with that said, this is an unnecessary BC break. Given that this behaviour has been known since PHP 4.2, I don't see why it is pressing enough to need to change it ASAP in the next minor version. So I voted -1 for it targeting PHP 7.2, but I'd be +1 on targeting the next major version instead.

pmmaga commented:
I have updated the RFC to target 8.0 and updated this PR to throw a `E_DEPRECATED` notice on the cases where it will change but keep the old behavior.

pmmaga commented:
Following the vote on https://wiki.php.net/rfc/negative_array_index and given that the deprecation notice was rejected, I'm closing this PR.

carusogabriel commented:
For reference: this was implemented in PHP 8.0 as #3772.
